crates/core/tedge_api/goals.md

```diff
@@ -0,0 +1,363 @@
+amount of different message kinds is limited and well known per-version. The
+message kinds are forward compatible. Meaning that new kinds of messages may be
+received but simply rejected if unknown.
+
+Messages conversations are also asynchronous, meaning that upon receiving a
+message ThinEdge might not reply. (This does not preclude the transport layer
+to assure that the message has been well received)
+
+This has two reasons:
+
+- It makes communication easier to implement
+- It reflects the network situation
+
+Using such an interface is still clunky without some additional help.
+
+As all messages have an ID, it could be possible to design a way of 'awaiting'
+a response. This is left as a future addition.
+
+## Plugins
+
+_Note: 'Plugins' is simply a working name for now. We are free to rename this
+to 'Extension' or any other name we think is good._
+
+Plugins are what makes ThinEdge work! They receive messages and execute their
+specific action if applicable.
+
+This part is split into two parts: PluginBuilders and Plugins themselves.
+
+### PluginBuilders
+
+A `PluginBuilder` is a Rust struct that implements `PluginBuilder`. They are the
+sole sources of plugin kinds in ThinEdge. (This is not a limitation for
+proprietary and non-rust plugins, read further to see how those are handled.)
+They are registered at startup and hardcoded into the ThinEdge binary.
+
+### Plugins
+
+A Plugin is a Rust struct that implements `Plugin`. It is clear that requiring
+all users who wish to extend ThinEdge to learn and use Rust is a _bad idea_. To
+make sure those users also get a first class experience special kinds of
+plugins are bundled per default: `stdio-external`, `http-external`. (The
+specific list will surely evolve over time.) These get instantiated like any
+other with the exception that it is expected that all messages are forwarded,
+e.g. to another binary or http endpoint. This way any user simply needs to know
+what kind of messages exists and how to send/receive them, giving them the same
+situation like a pure-Rust plugin would have.
+
+# Drawbacks
+
+- Tying the software specification to Rust code makes it more brittle to
+  accidental incompatibilities
+- People might get confused as to how Plugins can be written -> They see "Rust"
+  and think they need to know the language to make their own
+- Having everything in a single process group can make it harder to correctly
+  segment security rules for individual plugins
+
+# Rationale and alternatives
+
+Playing to Rust's strengths makes for better maintainable software. Using
+Traits to specify behaviour and "plain old Rust objects" also make the messages
+clear. As ThinEdge is a specialized piece of software, this should be embraced
+to deliver on an amazing experience.
+
+Alternatives could include:
+
+- Defining the objects externally, with for example CapNProto
+- Defining the objects through an API like Swagger or similar
+- Not doing this
+
+# Unresolved questions
```

Comment:
I would add questions

* Fine grain versus coarse grain for service definition, configuration and messages.
   * For instance, `ExecuteCommand{path, args, user}` or `RestartServiceX()`.
* How to configure the pre-requisite notably certificate creation, device registration.
   * Currently we are using different `tedge` commands for that and I don't see how this can be adapted/replaced.
* What methodology to define a meaning set of messages?
   * I propose to sketch plugin+message workflows for different representative use-cases of what has *already* been implemented with thin-edge.